AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_TextGeneration_with_LSTM/TextGenerationModelTraining.ipynb

```diff
@@ -0,0 +1,442 @@
+    "\n",
+    "After creating the model and adding Embedding layer, optimized LSTM layer from Itex can be added. Note, however, that this is sample for mentioned input text. Parameters are open to experiment, depending on the input text. Model accuracy with existing parameters is reaching 80%."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9758c3d6-9f7e-4ab6-9366-76ca1d23e63c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import intel_extension_for_tensorflow as itex"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "56d4d5c3-e2d1-4d13-9a5d-27731bf33206",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "neuron_coef = 4\n",
+    "itex_lstm_model = Sequential()\n",
+    "itex_lstm_model.add(Embedding(input_dim=vocab_size, output_dim=seq_length, input_length=seq_length))\n",
+    "itex_lstm_model.add(itex.ops.ItexLSTM(seq_length * neuron_coef, return_sequences=True))\n",
+    "itex_lstm_model.add(itex.ops.ItexLSTM(seq_length * neuron_coef))\n",
+    "itex_lstm_model.add(Dense(units=seq_length * neuron_coef, activation='relu'))\n",
+    "itex_lstm_model.add(Dense(units=vocab_size, activation='softmax'))\n",
+    "itex_lstm_model.summary()\n",
+    "itex_lstm_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])\n",
+    "itex_lstm_model.fit(x,y, batch_size=256, epochs=200)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e733a15f-e5f4-43f6-8f4f-ff6cabc5f374",
+   "metadata": {},
+   "source": [
+    "## Compared to LSTM from Keras\n",
+    "\n",
+    "The training done with Itex LSTM has efficient memory management on Intel GPU. As a reference, on the system with Intel® Arc™ 770, GPU memory was constant at around 4.5GB.\n",
+    "\n",
+    "Below is the example cell with the same dataset using LSTM layer from keras. To run on the same system, parameters such as sequence length, number of epochs, and other training layer parameters had to be lowered.\n",
+    "\n",
+    "Compared to parameters that were used by training with Itex LSTM (3192280 total parameters), with keras LSTM only 221870 total parameters were used. Besides accelerating the model training, Itex LSTM offers better memory management in Intel platform."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9a56430e-b17d-46d7-afad-30de62e71125",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from tensorflow.keras.layers import LSTM\n",
+    "\n",
+    "# Reducing the sequence to 10 compared to 50 with Itex LSTM\n",
+    "train_data_width = 10\n",
+    "tokens[:train_data_width]\n",
+    "\n",
+    "length = train_data_width + 1\n",
+    "lines = []\n",
+    "\n",
+    "for i in range(length, len(tokens)): \n",
+    "    seq = tokens[i - length:i]\n",
+    "    line = ' '.join(seq)\n",
+    "    lines.append(line)\n",
+    "\n",
+    "# Tokenization\n",
```

Comment:
Could we put tokenization codes into a function, and just call a function here instead? those codes are little distracted for itex and keras comparison

Reply:
Implemented with latest commit.